### PR TITLE
Handle app process crash.

### PR DESCRIPTION
**Pull request by dmitry-novikov**
Fixes issue when `Composer` fails to recognize that app process under the test is crashed that leads to successful completion.

### CODE REVIEW

File: composer/src/main/kotlin/com/gojuno/composer/Instrumentation.kt [hunk 1]
```diff
@@ -133,27 +137,29 @@ fun Observable<InstrumentationEntry>.asTests(): Observable<InstrumentationTest>
...
+                                        first.current == it.current &&
+                                        first.statusCode != it.statusCode
                                     }
 
                             if (second == null) null else first to second
                         }
                         .filterNotNull()
                         .map { (first, second) ->
                             InstrumentationTest(
+                                    index = first.current,
```
Comment (dmitry-novikov) [inline composer/src/main/kotlin/com/gojuno/composer/Instrumentation.kt]: Printing test index and total test count sharded to particular device really helps.

File: composer/src/main/kotlin/com/gojuno/composer/Instrumentation.kt [hunk 2]
```diff
@@ -54,19 +56,26 @@ private fun String.substringBetween(first: String, second: String): String {
...
         .trim()
 
-private fun parseUnableToFindInstrumentationInfo(str: String, output: File): Exception? = when (str.contains("INSTRUMENTATION_STATUS: Error=Unable to find instrumentation info for")) {
-    false -> null
-    true -> {
-        val runner = str.substringBetween("ComponentInfo{", "}").substringAfter("/")
-        Exception("Instrumentation was unable to run tests using runner $runner.\n" +
+private fun String.throwIfError(output: File) = when {
+    contains("INSTRUMENTATION_RESULT: shortMsg=Process crashed") -> {
+        throw Exception("Application process crashed. Check Logcat output for more details.")
```
Comment (artem-zinnatullin) [inline composer/src/main/kotlin/com/gojuno/composer/Instrumentation.kt]: nice

File: composer/src/main/kotlin/com/gojuno/composer/Instrumentation.kt [hunk 3]
```diff
@@ -99,8 +108,11 @@ fun readInstrumentationOutput(output: File): Observable<InstrumentationEntry> {
 
     return tail(output)
             .map(String::trim)
-            // `INSTRUMENTATION_CODE: -1` is last line printed by instrumentation, even if 0 tests were run.
-            .takeWhile { !it.startsWith("INSTRUMENTATION_CODE") }
+            .map { it.throwIfError(output) }
+            .takeWhile {
+                // `INSTRUMENTATION_CODE: <code>` is the last line printed by instrumentation, even if 0 tests were run.
```
Comment (artem-zinnatullin) [inline composer/src/main/kotlin/com/gojuno/composer/Instrumentation.kt]: thx for fix

File: composer/src/main/kotlin/com/gojuno/composer/TestRun.kt [hunk 4]
```diff
@@ -63,61 +60,68 @@ fun AdbDevice.runTests(
...
-                ignoredCount = tests.count { it.status is InstrumentationTest.Status.Ignored },
-                failedCount = tests.count { it.status is InstrumentationTest.Status.Failed },
-                durationNanos = System.nanoTime() - startTimeNanos,
-                timestampMillis = System.currentTimeMillis(),
-                logcat = logcatFileForDevice(logsDir),
-                instrumentationOutput = instrumentationOutputFile
-        )
-    }
+                adbDevice.log(
+                        "Test ${test.index}/${test.total} $status in " +
```
Comment (artem-zinnatullin) [inline composer/src/main/kotlin/com/gojuno/composer/TestRun.kt]: Super nice, I actually wanted this in #82

**Comment by artem-zinnatullin** [comment]
@yunikkk can you ship a release with this please?

**Pull request merged**